### DISCUSSION TITLE
Sharing my C/C++ folding rules

### DISCUSSION THREAD

**Discussion by HolyBlackCat** [zokugun/vscode-explicit-folding#44] [Show and tell]
Hello! Not sure if this is the right place to do it, but I just wanted to share the C/C++ folding rules ended up using.

What is folded:
* `{...}`, `(...)`, `[...]`
* block comments
* preprocessor conditionals
* raw strings: `R"(...)"`

Folding doesn't happen inside of:
* line and block comments
* string and character literals
* raw strings

Known issues that I don't know how to fix:
* Line continuation (`\` at the end of line) is not handled correctly in string and character literals, line comments, and everywhere else.
* Unbalanced brackets after `#define` should be ignored, but they aren't.

Config:
```jsonc
"cpp": [
    // Last version can be found at:  https://github.com/HolyBlackCat/vscode-settings/blob/master/settings.json
    { // `//...` (not actually folded, but forces contents to be ignored)
        "beginRegex": "(?!)",
        "endRegex": "\\/\\/.*",
    },
    { // `/*...*/`
        "begin": "/*",
        "end": "*/",
        "nested": false,
    },
    { // R"(...)" (C++ only)
        "beginRegex": "\\b(?:L|u8|u|U|)R\"([^\\s\\(\\)\\\\]{0,16})\\(",
        "endRegex": "\\)\\1\"",
        "nested": false,
        "foldLastLine": false,
    },
    { // "..." (not actually folded, but forces contents to be ignored)
        "beginRegex": "(?!)",
        "endRegex": "\"(?:[^\"\\\\]|\\\\.)*\"",
    },
    { // '...' (not actually folded, but forces contents to be ignored)
        // Note that we check that there's no digit before the first `'`, to support digit separators.
        "beginRegex": "(?!)",
        "endRegex": "(?:(?<=u8)|(?<![0-9]))'(?:[^'\\\\]|\\\\.)*'",
    },
    { // {...}
        "begin": "{",
        "end": "}",
        "foldLastLine": false,
    },
    { // (...)
        "begin": "(",
        "end": ")",
        "foldLastLine": false,
    },
    { // [...]
        "begin": "[",
        "end": "]",
        "foldLastLine": false,
    },
    { // Preprocessor conditionals
        "beginRegex": "^\\s*#\\s*if(?:(?:n)?def)?\\b", // "#if", "#ifdef", "#ifndef".
        "middleRegex": "^\\s*#\\s*el(?:se|if)\\b", // "#else", "#elif"
        "endRegex": "^\\s*#\\s*endif\\b", // "#endif"
    },
],
```

**Comment by daiyam** [comment]
@HolyBlackCat thanks for sharing your rules.
Your usage of the extension is way beyond its original purpose (emacs folding). :open_mouth:

For `#define`, have you tried:
```
{
  "begin": "#define",
  "continuation": "\\"
}
```

The `"beginRegex": "(?!)"` is baffling me... I will have check it...

**Comment by HolyBlackCat** [comment]
@daiyam `(?!)` is a regex that never matches anything. I'm basically exploiting #20. :P

The `#define` is complicated, so I decided to just ignore it.
```cpp
#define A /* 
// this should be folded
#define B */

#define A {
// this shouldn't be folded
#define B }
```
The second case (unbalanced braces in `#define`) almost never appear in actual code, so it's not a big loss.
